### PR TITLE
fixes handling of fancy nested classes in runtime reflection

### PR DESCRIPTION
**Pull request by xeno-by**
Replaces the `jclazz.isMemberClass` check for whether we have an
inner/nested class with `jclazz.getEnclosingClass != null`, because there
exist classes produced by javac (see the attached jar file and the test log)
which have the following properties:
- They are nested within a parent class
- getEnclosingClass returns a non-null value
- isMemberClass returns false

Previously such classes were incorrectly treated as non-nested, were
incorrectly put into an enclosing package rather than an enclosing class,
and had their names trimmed in the process, leading to situations when
a package has multiple declarations with the same name. This is now fixed.

When changing the check, we need to be careful with interpretation of
what Class.getEnclosingXXX methods return. If getEnclosingClass produces
a non-null result, this doesn't mean that the class is inner or nested,
because getEnclosingClass is also not null for local classes (the ones
with getEnclosingMethod != null || getEnclosingConstructor != null).
This is expressed in the order of pattern match clauses in `sOwner`.

Now when the bug is fixed, I also revert b18a2f8798b2, restoring a very
important integrity check in runtime reflection, which I had to disable
a couple hours ago to fix a master breakage. More details at scala-internals:
https://groups.google.com/forum/#!topic/scala-internals/hcnUFk75MgQ

### CODE REVIEW

**Comment by xeno-by** [comment]
review @retronym @paulp

File: src/reflect/scala/reflect/runtime/JavaMirrors.scala [hunk 1]
```diff
@@ -777,33 +777,19 @@ private[reflect] trait JavaMirrors extends internal.SymbolTable with api.JavaUni
...
-      }
+    // @eb: a weird classloader might return a null package for something with a non-empty package name
+    // for example, http://groups.google.com/group/scala-internals/browse_thread/thread/7be09ff8f67a1e5c
+    // in that case we could invoke packageNameToScala(jPackageName) and, probably, be okay
+    // however, I think, it's better to blow up, since weirdness of the class loader might bite us elsewhere
+    // [martin] I think it's better to be forgiving here. Restoring packageNameToScala.
+    private def sOwner(jclazz: jClass[_]): Symbol = jclazz match {
+      case PrimitiveOrArray()           => ScalaPackageClass
+      case EnclosedInMethod(ometh)      => methodToScala(ometh)
+      case EnclosedInConstructor(octor) => constructorToScala(octor)
```
Comment (paulp) [inline src/reflect/scala/reflect/runtime/JavaMirrors.scala]: Naming has gone off the rails when you're landing on strings like "octor". I understand it is "original constructor" (at least I assume that's what the "o" is) but still. Why not just call it "m" or "c", there is no loss of information. It exacts a comprehension tax to roll out new names which have no significance.

**Comment by retronym** [comment]
Can you please include instructions on how to generate the binary?

**Comment by retronym** [comment]
And is this actually exclusively a javac 7 issue? What does javac 6 emit?

**Comment by paulp** [comment]
I think it isn't a java7 issue per se, but an issue with the second class showing up in a package which fed in a $1, and java7 contributed the second to java.lang.

**Comment by xeno-by** [comment]
@retronym Yes, it's exactly as @paulp described.

File: src/reflect/scala/reflect/runtime/JavaMirrors.scala (hunk 1 above)
Comment (xeno-by) [inline src/reflect/scala/reflect/runtime/JavaMirrors.scala]: Fixed

**Comment by retronym** [comment]
In that case, can we construct a regular test case?

**Comment by xeno-by** [comment]
@retronym 
1) I've just provided the instructions on how to regenerate the jar if needed.
2) My first thought was indeed to construct a regular *.java test case. But then I realized that we can't trust javac to always produce the same classfiles that caused the bug, so we'd better have the classfiles pickled and stashed to be with us unchanged.

**Comment by retronym** [comment]
If we're worried about different javac output, we could assert that in the test as well.

File: src/reflect/scala/reflect/runtime/JavaMirrors.scala [hunk 2]
```diff
@@ -777,33 +777,19 @@ private[reflect] trait JavaMirrors extends internal.SymbolTable with api.JavaUni
...
-      }
+    // @eb: a weird classloader might return a null package for something with a non-empty package name
+    // for example, http://groups.google.com/group/scala-internals/browse_thread/thread/7be09ff8f67a1e5c
+    // in that case we could invoke packageNameToScala(jPackageName) and, probably, be okay
+    // however, I think, it's better to blow up, since weirdness of the class loader might bite us elsewhere
+    // [martin] I think it's better to be forgiving here. Restoring packageNameToScala.
+    private def sOwner(jclazz: jClass[_]): Symbol = jclazz match {
+      case PrimitiveOrArray()            => ScalaPackageClass
+      case EnclosedInMethod(jowner)      => methodToScala(jowner)
+      case EnclosedInConstructor(jowner) => constructorToScala(jowner)
```
Comment (paulp) [inline src/reflect/scala/reflect/runtime/JavaMirrors.scala:788]: Oh, the "o" was "owning", not "original". That's obvious now, but a good example of why attaching one letter prefixes to already-abbreviated strings is not the clearest.

**Comment by xeno-by** [comment]
@retronym Fair enough. Done.

**Comment by retronym** [comment]
BTW, the reason that this bug might seem non deterministic is javac first looks around for an existing anonymous class to use as the tag in the signature.

```
cat sandbox/Test.java
package p;
public class Test {
  private class C {};
  void foo() { new Runnable() { public void run() {}}; new C(); }
}

 Compiled from "Test.java"
class p.Test$C extends java.lang.Object{
    final p.Test this$0;
    p.Test$C(p.Test, p.Test$1);
}

Compiled from "Test.java"
class p.Test$1 extends java.lang.Object implements java.lang.Runnable{
    final p.Test this$0;
    p.Test$1(p.Test);
    public void run();
}
```

vs

```
cat sandbox/Test.java
package p;
public class Test {
  private class C {};
  void foo() { new C(); }
}

javap 'p/Test$1'
Compiled from "Test.java"
class p.Test$1 extends java.lang.Object{
}

  ~/code/scala javap 'p/Test$C'
Compiled from "Test.java"
class p.Test$C extends java.lang.Object{
    final p.Test this$0;
    p.Test$C(p.Test, p.Test$1);
}
```

**Comment by retronym** [comment]
I've got to head out for a while, so will finish reviewing a bit later. Thanks for jumping on this so quickly!

**Comment by retronym** [comment]
Unrelated to this change, I noticed that we have a stray '$' prefixing inner Java class names. Static inner classes are okay. Is this a known bug?

```
package p;
public class Test {
  private class C {};
  public class D {};
  public static class E {};
  void foo() { new C(); }
}
```

```
scala> reflect.runtime.currentMirror.reflect(new p.Test).symbol.typeSignature.declarations
res1: reflect.runtime.universe.MemberScope = SynchronizedOps(class $D, object $D, class $C, object $C, method foo, constructor Test)

scala> reflect.runtime.currentMirror.reflect(new p.Test).symbol.asClass.companionSymbol.asModule.moduleClass.typeSignature.declarations
res4: reflect.runtime.universe.MemberScope = SynchronizedOps(class E, object E)
```

**Comment by xeno-by** [comment]
@retronym Noted: https://issues.scala-lang.org/browse/SI-7921.

**Comment by xeno-by** [comment]
Just discovered https://issues.scala-lang.org/browse/SI-7922, which is NOT fixed by this pull request, even though it's very similar.

**Comment by retronym** [comment]
LGTM.

**Pull request merged**